File: p081.py
# Based on https://blog.dreamshire.com/project-euler-81-solution/ and https://www.mathblog.dk/project-euler-81-find-the-minimal-path-sum-from-the-top-left-to-the-bottom-right-by-moving-right-and-down/
# This solution is easier to understand than my own original solution (see first commit of p081.py)
def get_minimal_path_sum(filename):
    with open(filename) as f:
        matrix = [list(map(int, line.split(','))) for line in f.readlines()]

    # Modify the matrix in-place, where each cell eventually contains the minimum path sum to reach that cell
    for i in range(len(matrix)):
        for j in range(len(matrix)):
            # For each position (i,j), add to its current value the minimum of the value above or the value to the left
            matrix[i][j] += \
                min(matrix[i-1][j], matrix[i][j-1]) if i * j > 0 else \
                matrix[i-1][j] if i else \
                matrix[i][j-1] if j else \
                0

    return matrix[-1][-1]

if __name__ == "__main__":
    solution = get_minimal_path_sum("p081.input")
    print(solution)

File: p081.rs
mod minimal_path_sum;
use minimal_path_sum::read_file;
use integer_sqrt::IntegerSquareRoot;

fn get_minimal_path_sum(file_name: &str) -> usize {
	let mut matrix = read_file(file_name);
	let n = matrix.len().integer_sqrt();

	for i in 1..n {
		matrix[i] += matrix[i-1];
	}
	for i in (n..matrix.len()).step_by(n) {
		matrix[i] += matrix[i-n];
		for j in (i+1)..(i+n) {
			matrix[j] += matrix[j-1].min(matrix[j-n]);
		}
	}
	matrix[matrix.len() - 1]
}

#[test]
fn test_get_minimal_path_sum() {
	let actual = get_minimal_path_sum("p081_example.input");
	assert_eq!(actual, 2427);
}

fn main() {
	let solution = get_minimal_path_sum("p081.input");
	println!("{}", solution);
}